core/src/main/scala/scala/pickling/internal/package.scala

```diff
@@ -24,10 +23,18 @@ package object internal {
 
     }
   }
-  private[this] var currentRuntimeVar = new AtomicReference[spi.PicklingRuntime](initDefaultRuntime)
-  def currentRuntime: spi.PicklingRuntime = currentRuntimeVar.get
-  // Here we inject a new runtime for usage.
-  def replaceRuntime(r: spi.PicklingRuntime): Unit = {
+  private[this] var currentRuntimeVar = new AtomicReference[PicklingRuntime](initDefaultRuntime)
+  def currentRuntime: PicklingRuntime = currentRuntimeVar.get
+
+  /** Replace the old [[PicklingRuntime]] keeping its state.
+    *
+    * Note that we don't do anything with the [[RefRegistry]] because in
+    * future versions we are going to change how cyclic references work.
+    */
+  def replaceRuntime(r: PicklingRuntime): Unit = {
+    if(r.picklers.isLookupEnabled) {
+      currentRuntime.picklers.dumpStateTo(r.picklers)
```

Comment:
True, I didn't even thought about thread safety.
  
As you have correctly pointed out, the real problem is that we're not pre-allocating the object `PicklingProtocol` before using it. I forgot to explain that, my apologies. However, we cannot rely on the user to do this, since even if the pickling protocol is imported in scope, it won't be preallocated. One has to explicitly pull something from the protocol, like `implicitly[Pickler[List[Int]]]`. Then, it gets initialized and the runtime changes.
  
Since we cannot rely on the user always doing that, I implemented this ugly solution. Yet it is the only one that will work no matter what the user decides to do. We should probably work more on the thread safety. Do you think it would be a good idea to mark this method as `synchronized`? I'd do it just to shape a more robust api and make explicit the fact that it's not thread-safe. Anyway, as the user will rarely use it, the `synchronized` overhead will be minimal.